### DISCUSSION TITLE
Change the width dynamically without closing the tree

### DISCUSSION THREAD

**Discussion by aminnairi** [nvim-tree/nvim-tree.lua#2595] [Q&A] [answered]
Hello everyone!

I would like to change the width of the tree view dynamically.

For that, I was hopping to create a Which Key keybind that would allow me to set the width based on my liking.

For instance, pressing some keybind would allow the tree to be displayed using 30 units, another would allow the tree to dynamically resize the tree based on the largest opened node in terms of width.

I know that you can set this in the `setup` function using `width = 30` for a static width and `width = {}` for a dynamic width.

However, I would like to set this dynamically based on a keybind as stated above.

I read in the documentation that it is possible to use the `setup` more than once. The first time set the options for the next tree opening, and the other times it re-renders the tree entirely in order to account for the new settings.

However, doing so also closes the tree. 

```lua
require("nvim-tree").setup({ view = { width = newvaluehere }})
```

Is it possible to only apply the setting to the current opened tree instead?

I tried accessing the properties of the table directly although this is not documented, but as I expected it does not work.

```lua
require("nvim-tree").view.width = 30
```

Is it possible to achieve this goal using `nvim-tree`? If not, are there any other plugins that would allow me to achieve this?

Thank you!

**Comment by alex-courtis** [comment, marked as answer]
See `:help :NvimTreeResize`

There's an event to go with it, however we need to add some API.

**Comment by aminnairi** [comment]
Hi @alex-courtis and thank you for you reply.

I looked at the documentation for `:help NvimTreeResize` and there is no mention of dynamic resize mentionned here.

I also tried the `:NvimTreeResize {}` command with the `{}` Lua table, but as I guessed, `{}` has not the same meaninig since this is VimScript and not Lua.

However, using `:NvimTreeResize 30` or some other scalar value works as expected.

This would be great as a feature of the `NvimTreeResize` Vim command, even better if there would be an API as you suggest. I see something similar to Telescope.

```lua
require("nvim-tree.builtin").resize_width({ dynamic = true }) -- For dynamic width
require("nvim-tree-builtin").resize_width({ size = 30 }) -- For hardcoded width
require("nvim-tree-builtin").resize_width({ offset = 30 }) -- For increased width
require("nvim-tree-builtin").resize_width({ offset = -30 }) -- Fordecreased width
```

Or something similar.

**Reply by alex-courtis** [reply]
I see. Dynamic setting of view.width is the feature you're after here.

You could use something hacky with a view.width to set and persist:

```lua
local nvim_tree_width = 15

local function get_nvim_tree_width()
  return nvim_tree_width
end

local function set_nvim_tree_width(width)
  nvim_tree_width = width
  vim.cmd.NvimTreeResize(width)
end

vim.keymap.set("n", "<space>w", function() set_nvim_tree_width(50) end,      { noremap = true })

require("nvim-tree").setup({
  view = {
    width = get_nvim_tree_width,
```

**Reply by aminnairi** [reply]
Ah! Interesting, I should have though about it, the width can be a function.

This means that, according to your code snippet, we can store a local reference to the width inside of the scope of our Lua configuration file and update it through a function.

And I guess the `NvimTreeResize` command allow NvimTree to update the UI according to what has been passed here.

Based on your explanations, I over-engineered this a little bit to allow increementing and decrementing using a copycat of [React's useState](https://react.dev/reference/react/useState).

```lua
local function use_nvim_tree_width(options)
  if type(options) ~= "table" then
    return
  end

  ---@type number
  local offset = options.offset


  if type(offset) ~= "number" then
    offset = 10
  end

  ---@type number
  local previous_width = 30

  ---@type number|table
  local width = 30

  local function get_width()
    return width
  end

  local function increment_width()
    if type(width) == "table" then
      width = previous_width
    else
      local new_width = width + offset

      previous_width = new_width
      width = new_width
    end

    vim.cmd.NvimTreeResize(width)
  end

  local function decrement_width()
    if type(width) == "table" then
      width = previous_width
    else
      local new_width = width - offset

      previous_width = new_width
      width = new_width
    end

    vim.cmd.NvimTreeResize(width)
  end

  local function dynamic_width()
    width = {}
    vim.cmd.NvimTreeResize(width)
  end

  return get_width, increment_width, decrement_width, dynamic_width
end
```

I can then use it inside of my `config` function using this code exerpt.

```lua
local get_width, increment_width, decrement_width, dynamic_width = use_nvim_tree_width({
  offset = 10
})
```

The goal is then to bind them to shortcuts using `which-key`.

```lua
    whichKey.register({
      ["<leader>n"] = {
        name = "NvimTree",
        w = {
          name = "NvimTree Width",
          i = {
            increment_width,
            "Increment NvimTree Width"
          },
          d = {
            decrement_width,
            "Increment NvimTree Width"
          },
          a = {
            dynamic_width,
            "Automatic NvimTree Width"
          }
        },
...
```

And this works well! Except for the dynamic width, but again, as I suspected, the Vim command cannot take a table. And the offset +/- does not allow setting a -1 value for a dynamic width.

It would be great to have a Lua API that would allow users to update the tree's width using a function.

```lua
local function use_nvim_tree_width(options)
  -- The beginning of the function

  ---@type table|number
  local width = 30

  local function dynamic_width()
    width = {}
    require("nvim-tree.api").resize(width)
  end

  -- the rest of the function
end
```

I think allowing for the API to expose a `resize` function would allow much more interesting behavior to be encoded in Lua rather than using Vim commands.

Thank you again for your time, I really appreciated you pointing me in the right direction for this one as I didn't even know this would be possible. And with this almost perfect solution I can already be able to resize the window according to my needs without having to resort too much on the Vim command (even though this is little bit too much code to be honest!).

Should I open a new feature request for that one?

**Comment by alex-courtis** [comment]
That's quite a solution!

Raised #2598 

As usual, [contributions](https://github.com/nvim-tree/nvim-tree.lua/blob/master/CONTRIBUTING.md) are most gratefully appreciated
